File: src/android/app/src/main/java/org/wikimedia/epc/Identifier.java
package org.wikimedia.epc;

import java.util.Random;

public class Identifier {
    private Long timestamp;
    private Integer random;
    private Short sequence;
    public Identifier() {
        Random r = new Random();
        this.timestamp = System.currentTimeMillis() / 1000;
        this.random = r.nextInt(65535);
        this.sequence = 0;
    }
    public void step() {
        this.sequence++;
    }
    public String toHex(String separator) {
        String id = Long.toHexString(this.timestamp);
        id += separator + String.format("%04x", this.random);
        id += separator + String.format("%04x", this.sequence);
        return id;
    }
    public String toHex() {
        return this.toHex("");
    }
    public int randomComponent() {
        return this.random.intValue();
    }
}

File: src/ios/EPC Test App/EPC Test App/EventPlatformClient/Identifier.swift
import Foundation

/* Based on the Event Platform Client - ID specification:
 * https://docs.google.com/document/d/1bfMV3SzLqIlXVTZzuy1GrwyfZ0FT68DeHZSxj1GU04M/edit#
 */

struct Identifier {
    private let timestamp = UInt32(Date().timeIntervalSince1970)
    private let random = UInt16(arc4random_uniform(65535))
    private var sequence = UInt16(0)
    public mutating func step() -> Void {
        self.sequence += 1
    }
    public func toHex(separator: String = "") -> String {
        var id: String = String(self.timestamp, radix: 16)
        id += separator + String(format: "%04x", self.random)
        id += separator + String(format: "%04x", self.sequence)
        return id
    }
    public func randomComponent() -> Int {
        return Int(self.random);
    }
}
